controllers/pvc_storageclass_test.go

```diff
@@ -0,0 +1,320 @@
+			MutatingWebhookConfigRefs: []types.NamespacedName{{Name: webhookMutatingConfigName}},
+		}
+
+		err := cert.NewManager(nil, certOpts)
+		Expect(err).NotTo(HaveOccurred(), "unable to provision certificate")
+
+		webhookInstallOptions := &testEnv.WebhookInstallOptions
+		mgr, err := ctrl.NewManager(cfg, ctrl.Options{
+			Host:               webhookInstallOptions.LocalServingHost,
+			Port:               webhookInstallOptions.LocalServingPort,
+			CertDir:            webhookInstallOptions.LocalServingCertDir,
+			LeaderElection:     false,
+			MetricsBindAddress: "0",
+		})
+		Expect(err).NotTo(HaveOccurred(), "failed to create manager")
+
+		decoder, err := admission.NewDecoder(mgr.GetScheme())
+		Expect(err).NotTo(HaveOccurred(), "failed to create decoder")
+
+		pvcMutator := pvcmutator.NewController(mgr.GetClient(), decoder, []pvcmutator.Mutator{
+			storageclass.NewAnnotationSetter(mgr.GetClient()),
+		})
+
+		mgr.GetWebhookServer().Register(webhookMutatePVCsPath, &webhook.Admission{Handler: pvcMutator})
+		Expect(err).NotTo(HaveOccurred(), "failed to setup controller")
+
+		for _, sc := range storageClasses {
+			sc := sc
+			Expect(k8sClient.Create(ctx, &sc)).Should(Succeed())
+		}
+
+		go func() {
+			err := mgr.Start(ctx)
+			Expect(err).NotTo(HaveOccurred(), "failed to start manager")
+		}()
+
+		// Wait for manager to be ready.
+		time.Sleep(managerWaitDuration)
+	})
+
+	AfterEach(func() {
+		for _, sc := range storageClasses {
+			sc := sc
+			Expect(k8sClient.Delete(ctx, &sc)).Should(Succeed())
+		}
+
+		cancel()
+	})
+}
+
+var _ = Describe("PVC StorageClass UID to annotation controller", func() {
+	ctx := context.Background()
+
+	Context("When there is not StorageClass", func() {
+		SetupPVCStorageClassAnnotationTest(ctx)
+
+		It("The pvc should not be created", func() {
+			By("Expecting the PVC has default StorageClass")
+			pvc := genPVC()
+
+			By("Creating the PVC")
+			Expect(k8sClient.Create(ctx, &pvc)).ShouldNot(Succeed())
+		})
+	})
+
+	Context("When the default is not StorageOS not given", func() {
+		defaultStorageClass := getStorageClass(foreignDefaultStorageClassName, true)
+
+		SetupPVCStorageClassAnnotationTest(ctx, defaultStorageClass)
+
+		It("The pvc should be created", func() {
+			pvc := genPVC()
+
+			By("Creating the PVC")
+			Expect(k8sClient.Create(ctx, &pvc)).Should(Succeed())
+			defer func() {
+				Expect(k8sClient.Delete(ctx, &pvc)).Should(Succeed())
+			}()
+
+			By("Expecting the PVC has to be unchanged")
+			Consistently(func() corev1.PersistentVolumeClaim {
+				got := corev1.PersistentVolumeClaim{}
+				err := k8sClient.Get(ctx, client.ObjectKeyFromObject(&pvc), &got)
+				if err != nil {
+					return corev1.PersistentVolumeClaim{}
+				}
+				return got
+			}, storageClassTimeout, storageClassInterval).Should(Equal(pvc))
+		})
+	})
+
+	Context("When the default is not StorageOS given is not StorageOS", func() {
```

Comment:
Since this and the other context messages in the test case above and below seem to be a bit complex with two variations that can form a test matrix:
- Default StorageClass
- PVC's StorageClass

Maybe the tests would be simpler if we have two `Describe` blocks for when the default SC is storageos SC and non-storageos SC. And within them, we can have cases for PVCs that use different SCs.

I hope that'll simplify the Describe/Context messages for each of the tests. In the above Context, I would try to change it to `When the default SC is not STOS SC and the given PVC uses non-STOS SC`. It's a really long description of the test.

Reply:
Originally i did those two Context-es and created multiple It-s, but i had issues with context.Contex and also other resource creation in BeforeEach. I thought I'm totally wrong and instead of refactoring the entire setup which now looks similar than keygen, i went on this way. But creating a second Describe makes much more sense. Because I also feel nausea if that long names ;) Thanks